R/paste_linter.R

```diff
@@ -0,0 +1,74 @@
+#' Raise lints for several common poor usages of paste()
+#'
+#' The following issues are linted by default by this linter
+#'   (and each can be turned off optionally):
+#'
+#'  1. Block usage of [paste()] with `sep = ""`. [paste0()] is a faster, more concise alternative.
+#'  2. Block usage of `paste()` or `paste0()` with `collapse = ", "`. [toString()] is a direct
+#'     wrapper for this, and alternatives like [glue::glue_collapse()] might give better messages for humans.
+#'
+#' @evalRd rd_tags("paste_linter")
+#' @param allow_empty_sep Logical, default `FALSE`. If `TRUE`, usage of
+#'   `paste()` with `sep = ""` is not linted.
+#' @param allow_to_string Logical, default `FALSE`. If `TRUE`, usage of
+#'   `paste()` and `paste0()` with `collapse = ", "` is not linted.
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+paste_linter <- function(allow_empty_sep = FALSE, allow_to_string = FALSE) {
+  Linter(function(source_file) {
+    if (length(source_file$xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    if (allow_empty_sep) {
+      empty_sep_lints <- NULL
+    } else {
+      # NB: string-length()=2 means '' or "" (*not* 'ab' or 'cd' which have length 4)
```

Comment:
maybe a helper `string_to_code` is in order? I think we don't really need to support `r"-----(asd)-----"`, so the code could be rather simple:
1. use `deparse(str)` to generate the correct escapes for double quotes
2. substitute \" by " and ' by \' and change the quotes to single quotes
3. use `paste0('r"(', str, ')"')` as a reasonable approximation to the raw string literal (bugged if str contains `)"`, bug fixable should we actually need it)

Reply:
filed #1034, I think it makes most sense to address this globally